src/Widgets/VPN/VPNMenuItem.vala

```diff
@@ -16,106 +16,135 @@
-
-        notify["state"].connect (update);
-        radio_button.notify["active"].connect (update);
-        radio_button.button_release_event.connect ((b, ev) => {
-            user_action ();
-            return false;
+        connect_button = new Gtk.Button ();
+        connect_button.valign = Gtk.Align.CENTER;
+        connect_button.label = _("Connect");
+        connect_button.clicked.connect (() => {
+            if (state == State.CONNECTED_VPN) {
+                disconnect_clicked ();
+            } else {
+                connect_clicked ();
+            }
         });
+        size_group.add_widget (connect_button);
+
+        var grid = new Gtk.Grid ();
+        grid.margin = 6;
+        grid.column_spacing = 6;
+        grid.orientation = Gtk.Orientation.HORIZONTAL;
+        grid.attach (overlay, 0, 0, 1, 2);
+        grid.attach (vpn_label, 1, 0, 1, 1);
+        grid.attach (state_label, 1, 1, 1, 1);
+        grid.attach (vpn_info_button, 2, 0, 1, 2);
+        grid.attach (connect_button, 3, 0, 1, 2);
 
+        notify["state"].connect (update);
         connection.changed.connect (update);
         update ();
-    }
-
-    /**
-     * Only used for an item which is not displayed: hacky way to have no radio button selected.
-     **/
-    public VPNMenuItem.blank () {
-        radio_button = new Gtk.RadioButton (null);
-    }
 
-    public void set_active (bool active) {
-        radio_button.set_active (active);
+        add (grid);
+        show_all ();
     }
 
     private void update () {
-        radio_button.label = connection.get_id ();
+        vpn_label.label = connection.get_id ();
 
         switch (state) {
-        case State.FAILED_VPN:
-            show_item (error_img);
-            break;
-        case State.CONNECTING_VPN:
-            show_item (spinner);
-            break;
-        default:
-            hide_icons ();
-            break;
+            case State.FAILED_VPN:
+                state_label_text = _("Failed");
+                vpn_state.icon_name = "user-busy";
+                connect_button.label = _("Connect");
+                connect_button.sensitive = true;
+                connect_button.get_style_context ().remove_class (Gtk.STYLE_CLASS_DESTRUCTIVE_ACTION);
+                break;
+            case State.CONNECTING_VPN:
+                state_label_text = _("Connecting");
+                vpn_state.icon_name = "user-away";
+                connect_button.sensitive = false;
+                connect_button.get_style_context ().remove_class (Gtk.STYLE_CLASS_DESTRUCTIVE_ACTION);
+                break;
+            case State.CONNECTED_VPN:
+                state_label_text = _("Connected");
+                vpn_state.icon_name = "user-available";
+                connect_button.label = _("Disconnect");
+                connect_button.sensitive = true;
+                connect_button.get_style_context ().add_class (Gtk.STYLE_CLASS_DESTRUCTIVE_ACTION);
```

Comment:
I'm not sure this should be `destructive-action` unless disconnecting deletes or forgets some data. Presumably disconnecting does not remove the connection information right?

Reply:
I agree with @arshubham: disconnecting a VPN is destructive and not reversible; you and your data are potentially exposed at that point.